ddtrace/tracer/textmap.go

```diff
@@ -487,3 +501,309 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
+// Currently, Go tracer doesn't support 128-bit traceIDs, so the full traceID (32 hex-encoded digits) must be
+// stored into a field that is accessible from the span’s context. TraceId will be parsed from the least significant 16
+// hex-encoded digits into a 64-bit number.
+func parseTraceparent(ctx *spanContext, header string) error {
+	nonWordRegex := "\\W_"
+	header = strings.ToLower(strings.Trim(header, "\t -"))
+	if len(header) == 0 {
+		return ErrSpanContextNotFound
+	}
+	if len(header) != 55 {
+		return ErrSpanContextCorrupted
+	}
+	parts := strings.Split(header, "-")
+	if len(parts) != 4 {
+		return ErrSpanContextCorrupted
+	}
+	version := strings.Trim(parts[0], nonWordRegex)
+	if len(version) != 2 {
+		return ErrSpanContextCorrupted
+	}
+	if v, err := strconv.ParseUint(version, 16, 64); err != nil || v == 255 {
+		return ErrSpanContextCorrupted
+	}
+	// parsing traceID
+	fullTraceID := strings.Trim(parts[1], nonWordRegex)
+	if len(fullTraceID) != 32 {
+		return ErrSpanContextCorrupted
+	}
+	// checking that the entire TraceID is a valid hex string
+	if ok, err := regexp.MatchString("^[a-f0-9]+$", fullTraceID); !ok || err != nil {
+		return ErrSpanContextCorrupted
+	}
+	var err error
+	if ctx.traceID, err = strconv.ParseUint(fullTraceID[16:], 16, 64); err != nil {
+		return ErrSpanContextCorrupted
+	}
+	if ctx.traceID == 0 {
+		if strings.Trim(fullTraceID[:16], "0") == "" {
+			return ErrSpanContextNotFound
+		}
+	}
+	// parsing spanID
+	spanID := strings.Trim(parts[2], nonWordRegex)
+	if len(spanID) != 16 {
+		return ErrSpanContextCorrupted
+	}
+	if ok, err := regexp.MatchString("[a-f0-9]+", spanID); !ok || err != nil {
+		return ErrSpanContextCorrupted
+	}
+	if ctx.spanID, err = strconv.ParseUint(spanID, 16, 64); err != nil || ctx.spanID == 0 {
+		return ErrSpanContextNotFound
+	}
+	// parsing flags
+	flags := parts[3]
+	f, err := strconv.ParseInt(flags, 16, 8)
+	if err != nil {
+		return ErrSpanContextCorrupted
+	}
+	// setting trace-id to be used for span context propagation
+	setPropagatingTag(ctx, w3cTraceIDTag, fullTraceID)
+	ctx.setSamplingPriority(int(f)&0x1, samplernames.Unknown)
+	return nil
+}
+
+// parseTracestate attempts to parse tracestateHeader which is a list
+// with zero or more comma-separated (,) list-members.
+// An example value would be: `vendorname1=opaqueValue1,vendorname2=opaqueValue2,dd=s:1;o:synthetics`,
+// Where `dd` list contains values that would be in x-datadog-tags as well as those needed for propagation information.
+// The keys to the “dd“ values have been shortened as follows to save space:
+// `sampling_priority` = `s`
+// `origin` = `o`
+// `_dd.p.` prefix = `t.`
+func parseTracestate(ctx *spanContext, header string) error {
+	// if multiple headers are present, they must be combined and stored
+	setPropagatingTag(ctx, tracestateHeader, header)
+	list := strings.Split(strings.Trim(header, "\t "), ",")
```

Comment:
We should probably return an error here if `len(list) > 32` since that appears to be the max per the RFC

Reply:
I don't see anything that would support that in the W3C recommendation. It says to truncate the entry only
```
In a situation where tracestate needs to be truncated due to size limitations, the vendor MUST truncate whole entries. Entries larger than 128 characters long SHOULD be removed first. Then entries SHOULD be removed starting from the end of tracestate. Note that other truncation strategies like safe list entries, blocked list entries, or size-based truncation MAY be used, but are highly discouraged. Those strategies decrease the interoperability of various tracing vendors.
```

Reply:
I was just taking this from the RFC where it says:

> Restrictions on the contents of the tracestate header include the following:
> - There can be a maximum of 32 list-members

Maybe it's fine though, since this assumes that the header is valid in this way. Probably fine to leave it as is then.